code/modules/power/supermatter/supermatter_process.dm

```diff
@@ -138,6 +139,75 @@
+	animate(filters[1], time = 10 SECONDS, offset = 10, loop=-1)
+	animate(time = 10 SECONDS, offset = 0, loop=-1)
+	animate(filters[1], time = 2 SECONDS, size = 80, loop=-1, flags = ANIMATION_PARALLEL)
+	animate(time = 2 SECONDS, size = 10, loop=-1, flags = ANIMATION_PARALLEL)
+	animated = TRUE
+
+/obj/machinery/power/supermatter_crystal/proc/check_special_delamination() // In priority of devastation.
+	if(cascade_initiated) // Supermatter cascade totter goal.
+		return "cascade"
+	if(combined_gas > MOLE_PENALTY_THRESHOLD)// Singularity
+		return "singularity"
+	if(power > POWER_PENALTY_THRESHOLD) // Tesla delamination
+		return "tesla"
+	else
+		return null
+
+/obj/machinery/power/supermatter_crystal/proc/handle_crystal_effects(tick)
+	if(!check_special_delamination())
+		if(power || damage && !final_countdown)
+			set_light((initial(light_range) + power/200), initial(light_power) + power/1000, (gasmix_power_ratio > 0.8 ? SUPERMATTER_RED : SUPERMATTER_COLOUR), TRUE)
+
+			filters = filter(type="rays", size = clamp(power/30, 0, 125), color = (gasmix_power_ratio > 0.8 ? SUPERMATTER_RED : SUPERMATTER_COLOUR), factor = clamp(damage/600, 1, 10), density = clamp(damage/10, 12, 100))
+
+			if(!animated)// If we did this every tick we'd be crashing peoples clients all day.
+				animate_the_sm()
+
+		if(damage && final_countdown) // Let's jump to a special effect if we can.
+			set_light(initial(light_range) + clamp(damage*power, 50, 500), 3, (gasmix_power_ratio > 0.8 ? SUPERMATTER_RED : SUPERMATTER_COLOUR), TRUE)
+			filters = filter(type="rays", size = clamp((damage/100)*power, 50, 125), color = (gasmix_power_ratio > 0.8 ? SUPERMATTER_RED : SUPERMATTER_COLOUR), factor = clamp(damage/300, 1, 30), density = clamp(damage/5, 12, 200))
+
+	switch(check_special_delamination())
+		if("cascade")
+			set_light(initial(light_range) + clamp(damage*power, 50, 500), 3, SUPERMATTER_CASCADE_COLOUR, TRUE)
+			filters = filter(type="rays", size = clamp((damage/100)*power, 50, 125), color = SUPERMATTER_CASCADE_COLOUR, factor = clamp(damage/300, 1, 30), density = clamp(damage/5, 12, 200))
+			return
+
+		if("singularity")
+			if(!warp)
+				warp = new(src)
+				vis_contents += warp
+				animate(warp, time = 20*0.1*3, transform = matrix().Scale(0.5,0.5))
+				animate(time = 20*0.1*7, transform = matrix())
+
+			if(DT_PROB(1, 5))
+				set_light(initial(light_range) + clamp(damage*power, 10, 50), 3, SUPERMATTER_SINGULARITY_COLOUR, TRUE)
+
+			filters = filter(type="rays", size = clamp((damage/100)*power, 50, 125), color = SUPERMATTER_SINGULARITY_COLOUR, factor = clamp(damage/300, 1, 30), density = clamp(damage/5, 12, 200))
+			return
+
+		if("tesla")
+			set_light(initial(light_range) + clamp(damage*power, 50, 500), 3, SUPERMATTER_TESLA_COLOUR, TRUE)
+			filters = filter(type="rays", size = clamp((damage/100)*power, 50, 125), color = SUPERMATTER_TESLA_COLOUR, factor = clamp(damage/300, 1, 30), density = clamp(damage/5, 12, 200))
+			return
```

Comment:
This is interesting
why a switch case like this?

Reply:
uh not sure, I think I didn't want to write an if else thing, but this works for the three possible outcomes and makes it easier to add more, since what the proc returns can only be one of three in priority of what the SM counts for it's explosion at the time of

Reply:
why not just take the conditionals you're putting in the proc, and do them here?